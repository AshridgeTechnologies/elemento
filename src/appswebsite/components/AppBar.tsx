import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}