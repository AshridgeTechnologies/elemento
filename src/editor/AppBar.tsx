import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16 }}>
                        Elemento Studio
                    </Typography>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}