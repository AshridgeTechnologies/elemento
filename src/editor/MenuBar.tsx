import * as React from 'react';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';

export default function MenuBar({children = [] }: {children?: React.ReactNode}) {
    return (
        <Box sx={{flexGrow: 1}}>
            <Toolbar variant="dense" sx={{borderBottom: '1px solid #ddd'}}>
                {children}
            </Toolbar>
        </Box>
    );
}