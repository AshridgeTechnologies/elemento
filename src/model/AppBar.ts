import Element from './Element'
import BaseElement, {propDef} from './BaseElement'
import {ComponentType, ElementType, PropertyDef, PropertyValueType} from './Types'
import WebAssetIcon from '@mui/icons-material/WebAsset';
import * as theElements from './elements'

type Properties = { title?: PropertyValueType<string> }

export default class AppBar extends BaseElement<Properties> implements Element {

    static get iconClass() { return WebAssetIcon }
    type(): ComponentType { return 'statelessUI' }
    isLayoutOnly() { return true }

    get title() { return this.properties.title }

    canContain(elementType: ElementType) {
        const parentType = theElements[elementType].parentType
        return parentType === this.kind || parentType === 'any'
    }

    get propertyDefs(): PropertyDef[] {
        return [
            propDef('title')
        ]
    }

    static get parentType(): ElementType | 'any' | null { return 'App' }

}