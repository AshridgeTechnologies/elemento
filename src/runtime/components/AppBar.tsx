import React from 'react';
import {PropVal, StylesPropVals, valueOfProps} from '../runtimeFunctions'
import {AppBar as MuiAppBar, Stack, Toolbar, Typography} from '@mui/material'
import {sxProps, typographyStyles} from './ComponentHelpers'
import {pick} from 'ramda'

type Properties = { path: string, title?: PropVal<string>, children?: any, show?: PropVal<boolean>,
    styles?: StylesPropVals
}

export default function AppBar({children, path, ...props}: Properties) {
    const propVals = valueOfProps(props)
    const {title, show, styles = {}} = propVals
    const typographySxProps = sxProps(pick(typographyStyles, styles))
    return <MuiAppBar position="static" id={path} sx={sxProps(styles, show)}>
            <Toolbar variant="dense">
                <Stack
                    direction="row"
                    justifyContent="flex-start"
                    alignItems="center"
                    spacing={2}
                    flexWrap='wrap'
                    useFlexGap
                    rowGap={0}
                >
                {title && <Typography variant="h6" component="div" marginRight={2} sx={typographySxProps}>
                    {title}
                </Typography>}
                    <Stack direction="row"
                           justifyContent="flex-start"
                           alignItems="center"
                           spacing={1.5}
                           flexWrap='wrap'
                           useFlexGap
                           rowGap={0}
                    >
                        {children}
                    </Stack>
                </Stack>
            </Toolbar>
        </MuiAppBar>
}
