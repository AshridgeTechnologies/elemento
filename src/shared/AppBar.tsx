import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from './UserMenu'

type Properties = {title: string}
export default function AppBar({title}: Properties) {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        {title}
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}