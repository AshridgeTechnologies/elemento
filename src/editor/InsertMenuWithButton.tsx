import * as React from 'react'
import Button from '@mui/material/Button'
import {InsertMenuItemsFn, OnInsertFnWithPositionFn} from './Types'
import {Alert, Popover} from '@mui/material'
import {InsertMenu} from './InsertMenu'
import {ElementId, ElementType, InsertPosition} from '../model/Types'

export default function InsertMenuWithButton({onInsert, insertMenuItems, targetItemId}: {
    onInsert: OnInsertFnWithPositionFn,
    insertMenuItems: InsertMenuItemsFn,
    targetItemId: ElementId}) {
    const [anchorEl, setAnchorEl] = React.useState<Element | null>(null)
    const open = Boolean(anchorEl)
    const handleClose = () => setAnchorEl(null)
    const handleButtonClick = (event: React.MouseEvent) => {setAnchorEl(event.currentTarget)}
    const handleInsert = (insertPosition: InsertPosition, targetItemId: ElementId, elementType: ElementType) => {
        onInsert(insertPosition, targetItemId, elementType)
        handleClose()
    }

    const buttonId = 'insertButton'
    return (
        <div>
            <Button
                id={buttonId}
                aria-controls="insertMenu"
                aria-haspopup="true"
                aria-expanded={open ? 'true' : undefined}
                onClick={handleButtonClick}
            >
                Insert
            </Button>
            <InsertMenu anchorEl={anchorEl}
                        anchorOrigin={{vertical: 'bottom', horizontal: 'left'}}
                        transformOrigin={{vertical: 45, horizontal: 'left'}}
                        open={open && !!targetItemId} onClose={handleClose} onInsert={handleInsert}
                        insertMenuItems={insertMenuItems} targetItemId={targetItemId} labelledBy={buttonId}/>

            <Popover
                id='insertWarning'
                data-testid="insertWarning"
                open={open && !targetItemId}
                anchorEl={anchorEl}
                onClose={handleClose}
                anchorOrigin={{vertical: 'bottom', horizontal: 'left',}}
            >
                <Alert severity="warning">Please select the item where you want to insert</Alert>
            </Popover>
        </div>
    )
}