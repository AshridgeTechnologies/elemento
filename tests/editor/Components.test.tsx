/**
 * @vitest-environment jsdom
 */

import { afterEach, beforeEach, afterAll, beforeAll, describe, expect, it, vi, test } from "vitest"
import React from 'react'
import AppBar from '../../src/appsShared/AppBar'
import MenuBar from '../../src/editor/MenuBar'
import InsertMenuWithButton from '../../src/editor/InsertMenuWithButton'
import {snapshotTest} from '../testutil/testHelpers'

// vi.mock("firebase/auth", () => ({
//     getAuth: vi.fn().mockImplementation( ()=> ({currentUser() { return {name: 'Jo'}} })),
//     onAuthStateChanged: vi.fn()
// }))
// vi.mock("firebase/app", () => ({
//     initializeApp: vi.fn(),
// }))
//
describe('Components', () => {
    snapshotTest(<AppBar title='The title'/>)
    snapshotTest(<MenuBar/>)
    snapshotTest(<InsertMenuWithButton onInsert={() => {} } insertMenuItems={() => ['Text', 'TextInput', 'NumberInput','SelectInput', 'TrueFalseInput', 'Button', 'Data', 'Page'] } targetItemId={'id1'}/>)
})
