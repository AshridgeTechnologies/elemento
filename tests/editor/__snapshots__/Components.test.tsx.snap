// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components AppBar has expected structure 1`] = `
<div
  className="MuiBox-root css-i9gxme"
>
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionSticky css-7e0s21-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-dense css-10ltm04-MuiToolbar-root"
    >
      <button
        aria-label="menu"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </button>
      <div
        className="MuiTypography-root MuiTypography-h6 css-157e98d-MuiTypography-root"
      >
        The title
      </div>
      <div>
        <button
          aria-controls="userMenu"
          aria-haspopup="true"
          aria-label="account of current user"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AccountCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </button>
      </div>
    </div>
  </header>
</div>
`;

exports[`Components InsertMenu has expected structure 1`] = `
<div>
  <button
    aria-controls="insertMenu"
    aria-haspopup="true"
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    id="insertButton"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Insert
  </button>
</div>
`;

exports[`Components MenuBar has expected structure 1`] = `
<div
  className="MuiBox-root css-i9gxme"
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-dense css-9b53aa-MuiToolbar-root"
  />
</div>
`;
